drivers/i2c/i2c_npcx_controller.c

```diff
@@ -0,0 +1,878 @@
+
+		/* Clear rx FIFO threshold and status bits */
+		i2c_ctrl_clear_fifo_status(dev);
+
+		/* Wait for STOP completed */
+		data->oper_state = NPCX_I2C_WAIT_STOP;
+	} else {
+		/* Disable i2c interrupt first */
+		i2c_ctrl_irq_enable(dev, 0);
+		data->oper_state = NPCX_I2C_READ_SUSPEND;
+	}
+
+	return i2c_ctrl_notify(dev, 0);
+}
+
+static int i2c_ctrl_proc_write_msg(const struct device *dev,
+							struct i2c_msg *msg)
+{
+	struct i2c_ctrl_data *const data = DRV_DATA(dev);
+
+	data->is_write = 1;
+	data->ptr_msg = msg->buf;
+	data->msg = msg;
+
+	if (data->oper_state == NPCX_I2C_IDLE) {
+		data->oper_state = NPCX_I2C_WAIT_START;
+		/* Issue a START, wait for transaction completed */
+		i2c_ctrl_start(dev);
+
+		return i2c_ctrl_wait_completion(dev);
+	} else if (data->oper_state == NPCX_I2C_WRITE_SUSPEND) {
+		data->oper_state = NPCX_I2C_WRITE_FIFO;
+		i2c_ctrl_irq_enable(dev, 1);
+
+		return i2c_ctrl_wait_completion(dev);
+	}
+
+	LOG_ERR("Unexpected state %d during writing i2c port%02x!",
+					data->oper_state, data->port);
+	data->trans_err = -EIO;
+	return data->trans_err;
+}
+
+static int i2c_ctrl_proc_read_msg(const struct device *dev, struct i2c_msg *msg)
+{
+	struct i2c_ctrl_data *const data = DRV_DATA(dev);
+
+	data->is_write = 0;
+	data->ptr_msg = msg->buf;
+	data->msg = msg;
+
+	if (data->oper_state == NPCX_I2C_IDLE) {
+		data->oper_state = NPCX_I2C_WAIT_START;
+		/* Issue a START, wait for transaction completed */
+		i2c_ctrl_start(dev);
+
+		return i2c_ctrl_wait_completion(dev);
+	} else if (data->oper_state == NPCX_I2C_WRITE_SUSPEND) {
+		data->oper_state = NPCX_I2C_WAIT_RESTART;
+		/* Issue a RESTART, wait for transaction completed */
+		i2c_ctrl_start(dev);
+		i2c_ctrl_irq_enable(dev, 1);
+
+		return i2c_ctrl_wait_completion(dev);
+	} else if (data->oper_state == NPCX_I2C_READ_SUSPEND) {
+		data->oper_state = NPCX_I2C_READ_FIFO;
+
+		/* Setup threshold of RX FIFO first */
+		i2c_ctrl_rx_fifo_setup_threshold_nack(dev, msg->len,
+				(msg->flags & I2C_MSG_STOP) != 0);
+
+		/* Release bus */
+		i2c_ctrl_hold_bus(dev, 0);
+
+		/* Enable i2c interrupt first */
+		i2c_ctrl_irq_enable(dev, 1);
+		return i2c_ctrl_wait_completion(dev);
+	}
+
+	LOG_ERR("Unexpected state  %d during reading i2c port%02x!",
+					data->oper_state, data->port);
+	data->trans_err = -EIO;
+	return data->trans_err;
+}
+
+/* I2C controller isr function */
+static void i2c_ctrl_isr(const struct device *dev)
+{
+	struct smb_fifo_reg *const inst_fifo = HAL_I2C_FIFO_INSTANCE(dev);
+	struct i2c_ctrl_data *const data = DRV_DATA(dev);
+	uint8_t status, tmp;
+
+	status = inst_fifo->SMBST & NPCX_VALID_SMBST_MASK;
+	LOG_DBG("status: %02x, %d", status, data->oper_state);
+
+	/* A 'Bus Error' has been identified */
+	if (IS_BIT_SET(status, NPCX_SMBST_BER)) {
+		/* Generate a STOP condition immediately */
+		i2c_ctrl_stop(dev);
+
+		/* Clear BER Bit */
+		inst_fifo->SMBST |= BIT(NPCX_SMBST_BER);
+
+		/* Make sure slave doesn't hold bus by dummy reading */
```

Comment:
Do you mean this CL mentioned in the commit message?(https://chromium-review.googlesource.com/c/chromiumos/platform/ec/+/2373336) If so, I prefer to remove "dummy" directly. "fake" might cause misunderstanding since it does read from our FIFO register..

Reply:
Done.